### PR TITLE
feat(packageRules): add merge confidence matcher

### LINKED ISSUE #20550: packageRules.matchConfidence
### What would you like Renovate to be able to do?

Support a new option in packageRules - matchConfidence

### If you have any ideas on how this should be implemented, please tell us here.

This should be experimental _and_ undocumented like `minimumConfidence` is today, because for now it's only for internal Mend testing with a non-production API which is likely to change.

Ideally the endpoint should be configurable via environment variable to enable testers to experiment.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
- Added a merge confidence `packageRules` matcher.
- The fallback value for `getMergeConfidenceLevel` now returns `undefined`.
   - Currently, `neutral` is returned.
   - This is necessary because otherwise, the newly added matcher will match against it (when there is no or an invalid token).
   - Current flows that depend on this default `neutral` value are unaffected due to the usage of a nullish operator with a `neutral` default value.
- The `LookupUpdate` type now has an additional new optional prop `mergeConfidence`.

#### Additionally,
At startup, we try to query the API if a token is provided. We will abort the run on `403/5xx/ECONNRESET/ETIMEDOUT` failures to avoid PRs being auto-closed/opened unexpectedly.

If, for some reason, the API fails mid-run, we will abort as well as per the above cases.

## Context
Closes #20550

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -2135,6 +2135,38 @@ For example to apply a special label for Major updates:
...
+  "packageRules": [
+    {
+      "matchConfidence": ["high", "very high"],
+      "groupName": "high merge confidence"
+    }
+  ]
+}
+```
+
+This feature requires an API token. Set it via Host rules under the `merge-confidence` HostType:
```
Comment (viceice) [inline docs/usage/configuration-options.md:2157]: we shouldn't say twice, the warning above already mentioned the token requirement.

we should also add a warning that this is currently in private beta testing, so we don't get requests how to get that token.

File: docs/usage/self-hosted-experimental.md [hunk 2]
```diff
@@ -76,6 +76,10 @@ When using `matchPackageNames` and `matchPackagePatterns` matchers:
 1. Renovate first tries to match against `depName`
 2. If `depName` doesn't match then Renovate tries to match against `packageName`
 
+## `RENOVATE_X_MERGE_CONFIDENCE_API_BASE_URL`
+
+If set, Renovate will query this api for Merge Confidence data.
```
Comment (viceice) [inline docs/usage/self-hosted-experimental.md:81]: same as above, private beta.

File: lib/config/types.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@ import type { LogLevel } from 'bunyan';
 import type { PlatformId } from '../constants';
 import type { HostRule } from '../types';
 import type { GitNoVerifyOption } from '../util/git/types';
+import type { MergeConfidence } from '../util/merge-confidence';
```
Comment (viceice) [inline lib/config/types.ts:5]: type should be moved to types folder

File: lib/modules/manager/types.ts [hunk 4]
```diff
@@ -94,6 +95,7 @@ export interface LookupUpdate {
   pendingVersions?: string[];
   newVersion?: string;
   updateType?: UpdateType;
+  mergeConfidenceLevel?: MergeConfidence;
```
Comment (viceice) [inline lib/modules/manager/types.ts:98]:
```suggestion
  mergeConfidenceLevel?: MergeConfidence | undefined;
```
needed for new typescript validation, fix others too

File: lib/util/merge-confidence/index.spec.ts [hunk 5]
```diff
@@ -40,157 +49,329 @@ describe('util/merge-confidence/index', () => {
...
+      hostType: 'merge-confidence',
+      token: 'some-token',
+    };
+
     beforeEach(() => {
-      hostRules.clear();
+      jest.resetAllMocks();
+      process.env = {
+        ...envOrg,
+        RENOVATE_X_MERGE_CONFIDENCE_API_BASE_URL: apiBaseUrl,
```
Comment (viceice) [inline lib/util/merge-confidence/index.spec.ts:63]: don't clone ENV, just set the value. jest will isolate ENV for each test file

File: lib/util/merge-confidence/index.spec.ts [hunk 6]
```diff
@@ -40,157 +49,329 @@ describe('util/merge-confidence/index', () => {
...
-          'npm',
-          'renovate',
-          '24.1.0',
-          '25.0.0',
-          'bump'
-        )
-      ).toBe('neutral');
-    });
+    describe('getMergeConfidenceLevel()', () => {
+      it('returns neutral if undefined updateType', async () => {
```
Comment (viceice) [inline lib/util/merge-confidence/index.spec.ts:77]: please do required refactoring in separate PR, so we've less changes here

**Review by viceice — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md [hunk 7]
```diff
@@ -2135,6 +2135,38 @@ For example to apply a special label for Major updates:
 }
 ```
 
+### matchConfidence
+
+<!-- prettier-ignore -->
+!!! warning
+This configuration option is dependent on an API key, therefore it is effectively disabled if none is set.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:2142]:
```suggestion
This configuration option is dependent on a Mend API key, and is in private beta testing only. API keys are not available for free or via this repository.
```

Comment (rarkins) [reply docs/usage/configuration-options.md:2142]: @viceice like this?

File: docs/usage/configuration-options.md (hunk 1 above)
Comment (rarkins) [inline docs/usage/configuration-options.md:2157]:
```suggestion
Tokens can be configured via `hostRules` using the `merge-confidence` `hostType`:
```

File: docs/usage/self-hosted-experimental.md (hunk 2 above)
Comment (rarkins) [inline docs/usage/self-hosted-experimental.md:81]:
```suggestion
If set, Renovate will query this api for Merge Confidence data.
This feature is in private beta.
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 7 above)
Comment (viceice) [reply docs/usage/configuration-options.md:2142]: yes

File: docs/usage/configuration-options.md [hunk 8]
```diff
@@ -2135,6 +2135,38 @@ For example to apply a special label for Major updates:
 }
 ```
 
+### matchConfidence
+
+<!-- prettier-ignore -->
+!!! warning
+This configuration option is dependent on a Mend API key, and is in private beta testing only. API keys are not available for free or via this repository.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2142]:
```suggestion
!!! warning
    This configuration option needs a Mend API key, and is in private beta testing only.
    API keys are not available for free or via the `renovatebot/renovate` repository.
```

The formatting was nearly right, you need 4 spaces on each line. 😉 While I'm at it, a small rewrite to clarify.

File: docs/usage/configuration-options.md [hunk 9]
```diff
@@ -2135,6 +2135,38 @@ For example to apply a special label for Major updates:
...
+  "packageRules": [
+    {
+      "matchConfidence": ["high", "very high"],
+      "groupName": "high merge confidence"
+    }
+  ]
+}
+```
+
+Tokens can be configured via `hostRules` using the `merge-confidence` `hostType`:
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2157]:
```suggestion
Tokens can be configured via `hostRules` using the `"merge-confidence"` `hostType`:
```

I think we need to match the `"` in the JSON example here?

File: docs/usage/self-hosted-experimental.md (hunk 2 above)
Comment (HonkingGoose) [reply docs/usage/self-hosted-experimental.md:81]: Abbreviations like API should be capitalized. 😉

File: lib/config/options/index.ts [hunk 10]
```diff
@@ -1224,6 +1224,21 @@ const options: RenovateOptions[] = [
...
+    name: 'matchConfidence',
+    description:
+      'Merge confidence levels to match against (`low`, `neutral`, `high`, `very high`). Valid only within `packageRules` object.',
+    type: 'array',
+    subType: 'string',
+    allowedValues: ['low', 'neutral', 'high', 'very high'],
+    allowString: true,
+    stage: 'package',
+    parent: 'packageRules',
+    mergeable: true,
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:1237]: There's no `default: false` field for the `matchConfidence` option. Is that intended?

File: lib/util/merge-confidence/index.ts [hunk 11]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+ * @param datasource
+ * @param depName
+ * @param currentVersion
+ * @param newVersion
+ *
+ * @returns The merge confidence level for the given package release.
+ * @throws {ExternalHostError} if a timeout or connection reset error, authentication failure, or internal server error occurs during the request.
+ *
+ * @remarks
+ * Results are caches for 60 minutes to reduce the number of API calls.
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:117]:
```suggestion
 * Results are cached for 60 minutes to reduce the number of API calls.
```

File: lib/util/merge-confidence/index.ts [hunk 12]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
-  await packageCache.set('merge-confidence', token + url, confidence, 60);
+
+  await packageCache.set(hostType, cacheKey, confidence, 60);
   return confidence;
 }
+
+/**
+ * Checks the health of the Merge Confidence API by attempting to authenticate with it.
+ *
+ * @returns Resolves when the API health check is complete successfully.
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:160]:
```suggestion
 * @returns Resolves when the API health check is completed successfully.
```

File: lib/util/merge-confidence/index.ts [hunk 13]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+ *
+ * @throws {ExternalHostError} if a timeout, connection reset error, authentication failure, or internal server error occurs during the request.
+ *
+ * @remarks
+ * This function first checks that the API base URL and an authentication bearer token are defined before attempting to
+ * authenticate with the API. If either the base URL or token is no defined, it will immediately return
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:166]:
```suggestion
 * authenticate with the API. If either the base URL or token is not defined, it will immediately return
```

File: lib/util/merge-confidence/index.ts [hunk 14]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+  }
+
+  const url = `${apiBaseUrl}api/mc/availability`;
+  try {
+    await http.get(url);
+  } catch (err) {
+    apiErrorHandler(err);
+  }
+
+  logger.debug('merge confidence api - successfully authenticated');
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:184]:
```suggestion
  logger.debug('merge confidence API - successfully authenticated');
```

File: lib/util/merge-confidence/index.ts [hunk 15]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+ * @remarks
+ * This function first checks that the API base URL and an authentication bearer token are defined before attempting to
+ * authenticate with the API. If either the base URL or token is no defined, it will immediately return
+ * without making a request.
+ */
+export async function initMergeConfidence(): Promise<void> {
+  initConfig();
+
+  if (is.nullOrUndefined(apiBaseUrl) || is.nullOrUndefined(token)) {
+    logger.trace('merge confidence api usage is disabled');
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:173]:
```suggestion
    logger.trace('merge confidence API usage is disabled');
```

File: lib/util/merge-confidence/index.ts [hunk 16]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+  logger.debug('merge confidence api - successfully authenticated');
+  return;
+}
+
+function getApiBaseUrl(): string {
+  const defaultBaseUrl = 'https://badges.renovateapi.com/';
+  const baseFromEnv = process.env.RENOVATE_X_MERGE_CONFIDENCE_API_BASE_URL;
+
+  if (is.nullOrUndefined(baseFromEnv)) {
+    logger.trace('using default merge confidence api base url');
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:193]:
```suggestion
    logger.trace('using default merge confidence API base URL');
```

File: lib/util/merge-confidence/index.ts [hunk 17]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+    return defaultBaseUrl;
+  }
+
+  try {
+    const parsedBaseUrl = new URL(baseFromEnv).toString();
+    logger.trace(
+      { baseUrl: parsedBaseUrl },
+      'using merge confidence api base found in environment variables'
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:201]:
```suggestion
      'using merge confidence API base found in environment variables'
```

File: lib/util/merge-confidence/index.ts [hunk 18]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+    );
+    return parsedBaseUrl;
+  } catch (err) {
+    logger.warn(
+      { err, baseFromEnv },
+      'invalid merge confidence api base url found in environment variables - using default value instead'
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:207]:
```suggestion
      'invalid merge confidence API base URL found in environment variables - using default value instead'
```

File: lib/util/merge-confidence/index.ts [hunk 19]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+
+/**
+ * Handles errors returned by the Merge Confidence API.
+ *
+ * @param err - The error object returned by the API.
+ * @throws {ExternalHostError} if a timeout or connection reset error, authentication failure, or internal server error occurs during the request.
+ */
+function apiErrorHandler(err: any): void {
+  if (err.code === 'ETIMEDOUT' || err.code === 'ECONNRESET') {
+    logger.error({ err }, 'merge confidence api request failed - aborting run');
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:227]:
```suggestion
    logger.error({ err }, 'merge confidence API request failed - aborting run');
```

File: lib/util/merge-confidence/index.ts [hunk 20]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+    throw new ExternalHostError(err, hostType);
+  }
+
+  if (err.statusCode === 403) {
+    logger.error({ err }, 'merge confidence api token rejected - aborting run');
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:232]:
```suggestion
    logger.error({ err }, 'merge confidence API token rejected - aborting run');
```

File: lib/util/merge-confidence/index.ts [hunk 21]
```diff
@@ -44,51 +63,180 @@ const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence | null> =
...
+    throw new ExternalHostError(err, hostType);
+  }
+
+  if (err.statusCode >= 500 && err.statusCode < 600) {
+    logger.error({ err }, 'merge confidence api failure: 5xx - aborting run');
```
Comment (HonkingGoose) [inline lib/util/merge-confidence/index.ts:237]:
```suggestion
    logger.error({ err }, 'merge confidence API failure: 5xx - aborting run');
```

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 22]
```diff
@@ -1979,5 +1985,104 @@ describe('workers/repository/process/lookup/index', () => {
...
+
+        expect(getMergeConfidenceSpy).toHaveBeenCalledTimes(0);
+        expect(lookupUpdates).not.toMatchObject([
+          {
+            mergeConfidenceLevel: expect.anything(),
+          },
+        ]);
+      });
+
+      it('doesnt set merge confidence value when api is not in use', async () => {
```
Comment (HonkingGoose) [inline lib/workers/repository/process/lookup/index.spec.ts:2065]:
```suggestion
      it('doesn't set merge confidence value when API is not in use', async () => {
```

Or maybe you to to escape the `'` in `doesn't`?

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 23]
```diff
@@ -1979,5 +1985,104 @@ describe('workers/repository/process/lookup/index', () => {
...
+        const lookupUpdates = (await lookup.lookupUpdates(config)).updates;
+
+        expect(lookupUpdates).toMatchObject([
+          {
+            mergeConfidenceLevel: `high`,
+          },
+        ]);
+      });
+
+      it('doesnt get a merge confidence level when no packageRule is set', async () => {
```
Comment (HonkingGoose) [inline lib/workers/repository/process/lookup/index.spec.ts:2046]:
```suggestion
      it('doesn't get a merge confidence level when no packageRule is set', async () => {
```

Or maybe you to to escape the `'` in `doesn't`?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/options/index.ts (hunk 10 above)
Comment (rarkins) [reply lib/config/options/index.ts:1237]: Yes

File: lib/util/merge-confidence/index.spec.ts (hunk 5 above)
Comment (rarkins) [reply lib/util/merge-confidence/index.spec.ts:63]: If I make these changes:

```diff
diff --git a/lib/util/merge-confidence/index.spec.ts b/lib/util/merge-confidence/index.spec.ts
index f3abfd4ed..f7585945d 100644
--- a/lib/util/merge-confidence/index.spec.ts
+++ b/lib/util/merge-confidence/index.spec.ts
@@ -50,7 +50,6 @@ describe('util/merge-confidence/index', () => {
   });
 
   describe('API calling functions', () => {
-    const envOrg: NodeJS.ProcessEnv = process.env;
     const hostRule: HostRule = {
       hostType: 'merge-confidence',
       token: 'some-token',
@@ -58,17 +57,13 @@ describe('util/merge-confidence/index', () => {
 
     beforeEach(() => {
       jest.resetAllMocks();
-      process.env = {
-        ...envOrg,
-        RENOVATE_X_MERGE_CONFIDENCE_API_BASE_URL: apiBaseUrl,
-      };
+      process.env.RENOVATE_X_MERGE_CONFIDENCE_API_BASE_URL = apiBaseUrl;
       hostRules.add(hostRule);
       initConfig();
       memCache.reset();
     });
 
     afterEach(() => {
-      process.env = envOrg;
       hostRules.clear();
       resetConfig();
     });
```

Then the last two tests fail with missing mocks:

```
    Missing mocks!
     *   GET https://www.baseurl.com/api/mc/availability
```

I have not been able to resolve this. Can we stick with the current approach?

**Comment by Gabriel-Ladzaretti** [comment]
deconflicting

**Review by rarkins — DISMISSED** [review]

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by rarkins — DISMISSED** [review]

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 24]
```diff
@@ -1979,5 +1985,104 @@ describe('workers/repository/process/lookup/index', () => {
...
+      const hostRule: HostRule = {
+        hostType: 'merge-confidence',
+        token: 'some-token',
+      };
+
+      beforeEach(() => {
+        process.env = {
+          ...envOrg,
+          RENOVATE_X_MERGE_CONFIDENCE_API_BASE_URL: apiBaseUrl,
+        };
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.spec.ts:2005]: don't overwrite full env, only set the required value.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/lookup/index.spec.ts [hunk 25]
```diff
@@ -1979,5 +1985,98 @@ describe('workers/repository/process/lookup/index', () => {
...
+      };
+
+      beforeEach(() => {
+        hostRules.add(hostRule);
+        initConfig();
+        memCache.reset();
+      });
+
+      afterEach(() => {
+        hostRules.clear();
```
Comment (viceice) [inline lib/workers/repository/process/lookup/index.spec.ts:2007]:
```suggestion
```
already in upper afterEach

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]
